chirp/drivers/tk11.py

```diff
@@ -0,0 +1,3297 @@
+    REBOOT_REQUEST = 0x5DD
+
+
+class ConnectionResponse:
+    """ConnectionResponse"""
+
+    def __init__(self, message_type: int, message_length: int, version: str):
+        self.message_type = message_type
+        self.message_length = message_length
+        self.version = version
+
+    def __str__(self):
+        """Convert representation to string"""
+        return (
+            f"ConnectionResponse | "
+            f"message type : {str(self.message_type)} "
+            f"({self.message_length} bytes) - "
+            f"{self.version}"
+        )
+
+    @staticmethod
+    def from_bytes(bytes_array):
+        """Convert bytes to ConnectionResponse"""
+        if len(bytes_array) < 20:
+            raise errors.InvalidDataError("Connect response invalid format")
+
+        unpacked = struct.unpack('<HH16s', bytes_array[:20])
+
+        return ConnectionResponse(
+            unpacked[0],
+            unpacked[1],
+            unpacked[2].split(b'\x00')[0].decode('ascii', errors='ignore')
+        )
+
+
+class ConnectionRequest:
+    """ConnectionRequest"""
+    MESSAGE_LENGTH = 0x4
+
+    @staticmethod
+    def to_bytes():
+        """Convert request to bytes"""
+        return struct.pack(
+            '<HHI',
+            MessageType.CONNECT_REQUEST.value,
+            ConnectionRequest.MESSAGE_LENGTH,
+            MAGIC_CODE
+        )
+
+
+class MessageHeader(ctypes.Structure):
+    """MessageHeader structure"""
+    _pack_ = 1
+    _fields_ = [
+        ("message_type", ctypes.c_uint16),
+        ("message_length", ctypes.c_uint16),
+    ]
+
+
+class WriteRequest(ctypes.Structure):
+    """WriteRequest structure"""
+    _pack_ = 1
+    _fields_ = [
+        ("header", MessageHeader),
+        ("address", ctypes.c_uint32),
+        ("data_length", ctypes.c_uint16),
+        ("advance_flag", ctypes.c_uint8),
+        ("rsv", ctypes.c_uint8),
+        ("magic_code", ctypes.c_uint32),
+    ]
+
+
+class RebootRequest(ctypes.Structure):
+    """RebootRequest structure"""
+    _pack_ = 1
+    _fields_ = [
+        ("header", MessageHeader)
+    ]
+
+
+def parse_str_from_tk11(data: arrayDataElement, default_str: str = "") -> str:
+    """Parse string from TK11 format to chirp format as str.
+    Remove empty characters."""
+    raw = data.get_raw()
+
+    if not raw or raw[0] == 0xFF:
+        return default_str
+
+    try:
+        return raw.rstrip(b' \x00').decode("ascii")
+    except UnicodeDecodeError:
+        return default_str
+
+
+def parse_str_to_tk11(data: str, length: int = 16) -> list[int]:
+    """Parse string from chirp format to TK11 format.
+    Add empty character until a specific length"""
+    return list(data.ljust(length, "\x00").encode("ascii"))
+
+
+def get_response_code(response) -> int | None:
+    """Get response code or None of response length is < 2"""
+    if len(response) < 2:
+        return None
+
+    return response[0] | response[1] << 8
+
+
+def crc16_xmodem(content: bytes) -> int:
```

Comment:
We just added another xmodem CRC table to the tree (in the ha1g driver) and we have other CRC implementations. I'd really like to not add another one.